### ISSUE TITLE
"case" delta expanding fixpoints with coq-8.6

### ISSUE THREAD

**Issue by chdoc** [math-comp/math-comp#101]
I just ported some development from coq-8.5pl3 to 8.6 and noticed the following change in behavior:
```coq
From mathcomp Require Import all_ssreflect.
Set Implicit Arguments.
Unset Printing Implicit Defensive.
Import Prenex Implicits.

Inductive term := TT | And of term & term.
Fixpoint interp s := match s with TT => True | And s t => interp s /\ interp t end.
Definition foo s := And s s.

Goal forall s, interp (foo s) -> True.
  move => s. case.
  (* coq-8.5pl3/mathcomp-1.6.1 :  interp s -> interp s -> True *)

  (* coq-8.6/mathcomp-1.6.1 *)
  (* 
  (fix interp (s0 : term) : Prop := match s0 with
                                    | TT => True
                                    | And s1 t => interp s1 /\ interp t
                                    end) s ->
  (fix interp (s0 : term) : Prop := match s0 with
                                    | TT => True
                                    | And s1 t => interp s1 /\ interp t
                                    end) s -> True
  *)
Abort.

(* Destruct does not unfold interp (neither in 8.5pl3 nor in 8.6) *)
Goal forall s, interp (foo s) -> True.
  move => s. destruct 1 as [A B].
Abort.
```
The new behaviour leads to a lot less readable goal states and causes subsequent rewrites to fail. The problem does not occur when using "simpl" before "case". I have not tested with 1.6.1, but I recall seeing this a while ago, likely with 8.5pl1 or even 8.4. At least the older parts of the development did not need to be adapted.

**Comment by chdoc** [comment]
I just checked, the issue persists with coq-8.6.1/mathcomp-1.6.1. Could someone please comment on whether this is a bug or working as designed?

**Comment by herbelin** [comment]
Hi @chdoc: Your question is about a tactic provided by the SSReflect plugin, and it is clearly of a different nature than the MathComp library (if they used to be in the same repository, it is because they originated from the same core of developers/contributors).

The SSReflect plugin being released from Coq 8.7 with the main Coq archive, it is expected that SSReflect bugs are now reported on the [bug tracker](http://coq.inria.fr/bugs) of Coq (`SSreflect` component), even if for 8.6 a backport to version of SSReflect hosted in the present repository shall still have to be considered.

Was the change of behavior only dependent of the version of Coq or also of the version of the SSReflect plugin? In the former case, it is probably that the SSReflect version of `case` called a reduction function of Coq which changed behavior between 8.5 and 8.6. But at the same time, you seem to indicate that the problem already occurred with 8.4, so I'am a bit confused.

Best,

PS: Note that the bug tracker itself is actually to be moved tomorrow to github issues (a lot of moves these days...).

**Comment by CohenCyril** [comment]
@chdoc was this propagated to co ?
